From a3450be98a707c15a6974ff8f4000ae8388e1225 Mon Sep 17 00:00:00 2001
From: Xiongfeng Wang <wangxiongfeng2@huawei.com>
Date: Fri, 10 Sep 2021 09:00:22 +0800
Subject: [PATCH 2882/4585] [Huawei] PCI: Add quirk for HiSilicon PCI NP
 devices SD581x

Offering: HULK
hulk inclusion
category: feature
bugzilla: 181824

--------------------------------

Add quirk for HiSilicon PCI network processor devices SD581x.
Correct the 'class' register and disable other BARs except BAR0.

Reviewed-by: Hanjun Guo <guohanjun@huawei.com>
Signed-off-by: Chen Jun <chenjun102@huawei.com>
Patchwork Links: http://patchwork.huawei.com/project/hulk5.10/list/?series=18741
Signed-off-by: Xiongfeng Wang <wangxiongfeng2@huawei.com>
---
 drivers/pci/quirks.c    | 14 ++++++++++++++
 include/linux/pci_ids.h |  3 +++
 2 files changed, 17 insertions(+)

diff --git a/drivers/pci/quirks.c b/drivers/pci/quirks.c
index 87e88e8f6a2c..26b526ed90a8 100644
--- a/drivers/pci/quirks.c
+++ b/drivers/pci/quirks.c
@@ -5822,8 +5822,16 @@ static void pci_quirk_hisi_fixup_class(struct pci_dev *dev)
 	pci_info(dev, "force hisi class type to network\n");
 }
 
+DECLARE_PCI_FIXUP_EARLY(PCI_VENDOR_ID_HUAWEI, PCIE_DEVICE_ID_HISI_504F,
+			pci_quirk_hisi_fixup_class);
+DECLARE_PCI_FIXUP_EARLY(PCI_VENDOR_ID_HUAWEI, PCIE_DEVICE_ID_HISI_5813,
+			pci_quirk_hisi_fixup_class);
+DECLARE_PCI_FIXUP_EARLY(PCI_VENDOR_ID_HUAWEI, PCIE_DEVICE_ID_HISI_5815,
+			pci_quirk_hisi_fixup_class);
 DECLARE_PCI_FIXUP_EARLY(PCI_VENDOR_ID_HUAWEI, PCIE_DEVICE_ID_HISI_5816,
 			pci_quirk_hisi_fixup_class);
+DECLARE_PCI_FIXUP_EARLY(PCI_VENDOR_ID_HUAWEI, PCIE_DEVICE_ID_HISI_5873,
+			pci_quirk_hisi_fixup_class);
 DECLARE_PCI_FIXUP_EARLY(PCI_VENDOR_ID_HUAWEI, PCIE_DEVICE_ID_HISI_5886,
 			pci_quirk_hisi_fixup_class);
 DECLARE_PCI_FIXUP_EARLY(PCI_VENDOR_ID_HUAWEI, PCIE_DEVICE_ID_HISI_5891,
@@ -5837,6 +5845,8 @@ static void pci_quirk_hisi_fixup_bar(struct pci_dev *dev)
 
 	if (dev->device == PCIE_DEVICE_ID_HISI_504F ||
 		dev->device == PCIE_DEVICE_ID_HISI_5806 ||
+		dev->device == PCIE_DEVICE_ID_HISI_5813 ||
+		dev->device == PCIE_DEVICE_ID_HISI_5815 ||
 		dev->device == PCIE_DEVICE_ID_HISI_5816 ||
 		dev->device == PCIE_DEVICE_ID_HISI_5891 ||
 		dev->device == PCIE_DEVICE_ID_HISI_5892 ||
@@ -5868,6 +5878,10 @@ DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_HUAWEI, PCIE_DEVICE_ID_HISI_504F,
 			 pci_quirk_hisi_fixup_bar);
 DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_HUAWEI, PCIE_DEVICE_ID_HISI_5806,
 			 pci_quirk_hisi_fixup_bar);
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_HUAWEI, PCIE_DEVICE_ID_HISI_5813,
+			 pci_quirk_hisi_fixup_bar);
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_HUAWEI, PCIE_DEVICE_ID_HISI_5815,
+			 pci_quirk_hisi_fixup_bar);
 DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_HUAWEI, PCIE_DEVICE_ID_HISI_5816,
 			 pci_quirk_hisi_fixup_bar);
 DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_HUAWEI, PCIE_DEVICE_ID_HISI_5886,
diff --git a/include/linux/pci_ids.h b/include/linux/pci_ids.h
index 611a13e14665..4c2c58f9d8fa 100644
--- a/include/linux/pci_ids.h
+++ b/include/linux/pci_ids.h
@@ -2572,7 +2572,10 @@
 /* Hisilicon PCIe NP devices */
 #define PCIE_DEVICE_ID_HISI_504F	0x504f
 #define PCIE_DEVICE_ID_HISI_5806	0x5806
+#define PCIE_DEVICE_ID_HISI_5813	0x5813
+#define PCIE_DEVICE_ID_HISI_5815	0x5815
 #define PCIE_DEVICE_ID_HISI_5816	0x5816
+#define PCIE_DEVICE_ID_HISI_5873	0x5873
 #define PCIE_DEVICE_ID_HISI_5886	0x5886
 #define PCIE_DEVICE_ID_HISI_5891	0x5891
 #define PCIE_DEVICE_ID_HISI_5892	0x5892
-- 
2.17.1

